runtime/tr.source/trj9/z/codegen/ReduceSynchronizedFieldLoad.cpp

```diff
@@ -66,71 +70,112 @@ ReduceSynchronizedFieldLoad::inlineSynchronizedFieldLoad(TR::Node* node, TR::Cod
+      bool loadRegisterRequiresShift = false;
 
-         generateSSFInstruction(cg, TR::InstOpCode::LPDG, node, registerPair, alignedLoadMemoryReference, lockMemoryReference);
+      bool lockRegisterRequires32BitLogicalSignExtension = false;
+      bool loadRegisterRequires32BitLogicalSignExtension = false;
 
-         // For 64-bit uncompressed lockwords, when the load memory reference is not aligned to double-word boundaries,
-         // the loaded value is at the lower 32-bit. We need an AND to zero out the high half of loadRegister.
-         if (node->getChild(1)->getOpCodeValue() == TR::iloadi)
+      if (is32BitLock)
+         {
+         if ((lockWordOffset & 7) == 0)
             {
-            generateRILInstruction(cg, TR::InstOpCode::NIHF, node, loadRegister, static_cast<uintptrj_t>(0));
+            lockRegisterRequiresShift = true;
+            }
+         else
+            {
+            // This is because we must use LPDG to load a 32-bit value using displacement -4
+            TR_ASSERT_SAFE_FATAL((lockWordOffset & 3) == 0, "Lockword must be aligned on a word boundary\n");
+
+            lockRegisterRequires32BitLogicalSignExtension = true;
+            alignedLockMemoryReference = generateS390MemoryReference(*lockMemoryReference, -4, cg);
             }
          }
       else
          {
-         TR::MemoryReference* reusedLoadMemoryReference = generateS390MemoryReference(*loadMemoryReference, 0, cg);
-
-         generateSSFInstruction(cg, TR::InstOpCode::LPDG, node, registerPair, reusedLoadMemoryReference, lockMemoryReference);
+         // This is because we must use LPDG to load a 64-bit value
+         TR_ASSERT_SAFE_FATAL((lockWordOffset & 7) == 0, "Lockword must be aligned on a double-word boundary\n");
+         }
 
-         // We know we are generating for a 64-bit target. Furthermore we know the field we are loading is on a double-
-         // word boundary. However if we are loading an integer it only occupies 32-bits of storage, and since we must
-         // use LPDG to load the field it means the field will be generated in the most significant 32-bits of the
-         // loaded register. Hence we must carry out a right shift to move the loaded value into the least significant
-         // 32-bits of the respective register.
-         if (node->getChild(1)->getOpCodeValue() == TR::iloadi)
+      if (is32BitLoad)
+         {
+         if ((loadMemoryReference->getOffset() & 7) == 0)
             {
-            generateRSInstruction(cg, TR::InstOpCode::SRLG, node, loadRegister, loadRegister, 32);
+            loadRegisterRequiresShift = true;
+            }
+         else
+            {
+            // This is because we must use LPDG to load a 32-bit value using displacement -4
+            TR_ASSERT_SAFE_FATAL((loadMemoryReference->getOffset() & 3) == 0, "Field must be aligned on a word boundary\n");
+
+            loadRegisterRequires32BitLogicalSignExtension = true;
```

Comment:
If we are loading an 32-bit int, do we still need to logical extend to zero out the upper 32-bits?

Reply:
Yes, because under compressedrefs shift of zero we need to load a full 64-bit uncompressed value. The `LPDG` will do a 32-bit load but we need to sign extend it. We behave similarly to what we would have generated for a normal load, which would have been a `LLGF`. In our case `LPDG` + `LLGFR`.

Reply:
Note we would not generate the `SLLG` below because it is guarded by `loadNode->getFirstChild()->getOpCodeValue() == TR::lshl`. This node does not exist for compressedrefs shift of zero.

Reply:
Yeah, I was thinking for primitive integer types the LLGFR is unnecessary.  However, I wonder how this interacts with our fancy load evaluator.  In some cases, we do extension as part of the load evaluation.

Reply:
I made the modification in 86c3047 to only generate the extension for address data types. This should address all the review comments. Let me know if there is anything else.